<!DOCTYPE html>
<html lang='en'>

<head>
  <meta charset="utf-8">
    <title>SHOOT FAQs</title>
    <!-- Latest compiled and minified CSS -->
    <script src="https://code.jquery.com/jquery.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/underscore.js/1.8.3/underscore-min.js"></script>


    <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
    <link href="//netdna.bootstrapcdn.com/font-awesome/4.0.3/css/font-awesome.css" rel="stylesheet">

    <!-- Optional theme -->
    <link rel="stylesheet" href="//stackpath.bootstrapcdn.com/bootswatch/4.1.1/spacelab/bootstrap.min.css">

    <!-- Latest compiled and minified JavaScript -->
    <script src="//cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js" integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q" crossorigin="anonymous"></script>
    <script src="//maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js" integrity="sha384-JZR6Spejh4U02d8jOt6vLEHfe/JQGiRRSQQxSfFWpi1MquVdAyjUar5+76PVCmYl" crossorigin="anonymous"></script>

    <style>
      nav {
        margin-bottom: 25px;
      }
    </style>
</head>

<body>

<link rel="shortcut icon" href="{{ url_for('static', filename='favicon.ico') }}">

<nav class="navbar navbar-expand-lg navbar-light bg-light">
  <a class="navbar-brand" href="/">SHOOT: sprout a new branch on the tree of life</a>
</nav>

<div class="container" id = 'faqs'>   
    <h2>FAQs</h2>
    <ol>
        <li>How are the gene trees inferred?</li>
        <ul>
            <li>Maximum likelihood tree inference from a multiple sequence alignment is used. 
The sequences for each database tree are aligned using MAFFT L-INS-i for up to 500 sequences 
or default MAFFT for larger alignments. Trees are inferred using FastTree with default parameters.
This is performed by OrthoFinder, which also roots the tree using a rooted gene tree to identify 
the outgroup genes. A user's query gene is added to the alignment using MAFFT with the '--add' functionality and
is added to the pre-computed gene tree using IQTREE's '-g' option. The user's tree is rooted
on the same genes as the database tree.</li>
        </ul>
        <li>How is the tree visualisation done?</li>
        <ul>
          <li>All credit is due to phlotree.js developed by Stephen D. Shank, Steven Weaver & Sergei L. Kosakovsky Pond (have I missed anyone?). It was very easy to use and looks great! The github page is here: <a href="https://github.com/veg/phylotree.js">https://github.com/veg/phylotree.js</a> and the citation is: 
          <ul style="list-style-type:none;">
            <li>Shank, S., Weaver, S. & Kosakovsky Pond, S. phylotree.js - a JavaScript library for application development and interactive data visualization in phylogenetics. BMC Bioinformatics 19, 276 (2018). <a href="https://doi.org/10.1186/s12859-018-2283-2">https://doi.org/10.1186/s12859-018-2283-2</a></li>
          </ul>
          </li>
        </ul>
        <li>How do we cite SHOOT?</li>
        <ul>
            <li>There will be a preprint in the next few weeks.</li>
        </ul>
        <li>What databases does SHOOT use?</li>
        <ul>
            <li>SHOOT will very soon offer a range of curated databases. The motivation for encouraging the use of curated databases over defaulting to the BLAST nr database is that these curated databases should make it easier to picture the evolution of the whole gene family rather than overwhelming the tree with thousands of near-identical hits. The current database is:</li>
            <ul>
                <li>SHOOT <a href="https://www.uniprot.org/help/reference_proteome">UniProt Reference Proteomes 2020</a>.</li>
                <p>This database contains 78 species across all domains of life. UniProt says: "Reference proteomes have been selected among all proteomes (manually and algorithmically, according to a number of criteria) to provide broad coverage of the tree of life."</p>
                <p>You can see the (partially resolved) species tree <a href="{{ url_for('SpeciesTree_UniProt') }}">here</a></p>
                <table class="table">
                    <thead>
                      <tr>
                        <th>Eukaryota</th>
                        <th>Bacteria</th>
                        <th>Archaea</th>
                      </tr>
                    </thead>
                    <tbody>
                      <tr>
                        <td>Anopheles gambiae</td>
                        <td>Aquifex aeolicus</td>
                        <td>Halobacterium salinarum</td>
                      </tr>
                      <tr>
                        <td>Arabidopsis thaliana</td>
                        <td>Bacillus subtilis</td>
                        <td>Korarchaeum cryptofilum</td>
                      </tr>
                      <tr>
                        <td>Batrachochytrium dendrobatidis</td>
                        <td>Bacteroides thetaiotaomicron</td>
                        <td>Methanocaldococcus jannaschii</td>
                      </tr>
                      <tr>
                        <td>Bos taurus</td>
                        <td>Bradyrhizobium diazoefficiens</td>
                        <td>Methanosarcina acetivorans</td>
                      </tr>
                      <tr>
                        <td>Branchiostoma floridae</td>
                        <td>Chlamydia trachomatis</td>
                        <td>Nitrosopumilus maritimus</td>
                      </tr>
                      <tr>
                        <td>Caenorhabditis elegans</td>
                        <td>Chloroflexus aurantiacus</td>
                        <td>Saccharolobus solfataricus</td>
                      </tr>
                      <tr>
                        <td>Candida albicans</td>
                        <td>Deinococcus radiodurans</td>
                        <td>Thermococcus kodakarensis</td>
                      </tr>
                      <tr>
                        <td>Canis lupus familiaris</td>
                        <td>Dictyoglomus turgidum</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Chlamydomonas reinhardtii</td>
                        <td>Escherichia coli</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Ciona intestinalis</td>
                        <td>Fusobacterium nucleatum</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Cryptococcus neoformans</td>
                        <td>Geobacter sulfurreducens</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Danio rerio</td>
                        <td>Gloeobacter violaceus</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Dictyostelium discoideum</td>
                        <td>Helicobacter pylori</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Drosophila melanogaster</td>
                        <td>Leptospira interrogans</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Gallus gallus</td>
                        <td>Mycobacterium tuberculosis</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Giardia intestinalis</td>
                        <td>Mycoplasma genitalium</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Gorilla gorilla gorilla</td>
                        <td>Neisseria meningitidis</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Helobdella robusta</td>
                        <td>Pseudomonas aeruginosa</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Homo sapiens</td>
                        <td>Rhodopirellula baltica</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Ixodes scapularis</td>
                        <td>Streptomyces coelicolor</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Leishmania major</td>
                        <td>Synechocystis sp.</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Lepisosteus oculatus</td>
                        <td>Thermodesulfovibrio yellowstonii</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Monodelphis domestica</td>
                        <td>Thermotoga maritima</td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Monosiga brevicollis</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Mus musculus</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Nematostella vectensis</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Neosartorya fumigata</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Neurospora crassa</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Oryza sativa subsp. japonica</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Oryzias latipes</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Pan troglodytes</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Paramecium tetraurelia</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Phaeosphaeria nodorum</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Physcomitrella patens</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Phytophthora ramorum</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Plasmodium falciparum</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Puccinia graminis</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Rattus norvegicus</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Saccharomyces cerevisiae</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Schizosaccharomyces pombe</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Sclerotinia sclerotiorum</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Thalassiosira pseudonana</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Tribolium castaneum</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Trichomonas vaginalis</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Ustilago maydis</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Xenopus tropicalis</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Yarrowia lipolytica</td>
                        <td></td>
                        <td></td>
                      </tr>
                      <tr>
                        <td>Zea mays</td>
                        <td></td>
                        <td></td>
                      </tr>
                    </tbody>
                    </table>
            </ul>
        </ul>
        <li>Could my tree need to be manually re-rooted?</li>
        <ul>
            <li>This is possible if your gene is an out-group to the species in the database
you searched against. It's not possible to detect this programmatically so you should consider
whether the tree should be re-rooted on your query gene. A far better approach is to search against
a database which has a taxonomic range which covers your species. SHOOT should then be able to root
the gene tree correctly. It isn't necessary for the database to actually contain your species.</li>
        </ul>
        <li>Can we put SHOOT as a gene search tool on our genome resource website?</li>
        <ul>
            <li>Yes! Please get in touch and I'll help you get set up.</li>
        </ul>
    </ol>
</div>

<footer class="footer fixed-bottom bg-light">
  <div class="container">
    <span class="text-muted">SHOOT: gene search and placement within a phylogenetic database. David Emms & Steven Kelly &copy; 2021</span>
  </div>
</footer>

</body>