<!DOCTYPE html>
<html lang='en'>

<head>
  <meta charset="utf-8">
    <title>SHOOT.bio Help & Documentation</title>
    <!-- Google Analytics -->
    <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
      
      ga('create', 'UA-153557091-2', 'auto');
      ga('send', 'pageview');
    </script>
    <!-- End Google Analytics -->

    <!-- Latest compiled and minified CSS -->
    <script src="https://code.jquery.com/jquery.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/underscore.js/1.8.3/underscore-min.js"></script>


    <!-- <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous"> -->
    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-EVSTQN3/azprG1Anm3QDgpJLIm9Nao0Yz1ztcQTwFspd3yD65VohhpuuCOmLASjC" crossorigin="anonymous">
    <link href="//netdna.bootstrapcdn.com/font-awesome/4.0.3/css/font-awesome.css" rel="stylesheet">

    <!-- Optional theme -->
    <link rel="stylesheet" href="//stackpath.bootstrapcdn.com/bootswatch/4.1.1/spacelab/bootstrap.min.css">

    <!-- Latest compiled and minified JavaScript -->
    <script src="//cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js" integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q" crossorigin="anonymous"></script>
    <script src="//maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js" integrity="sha384-JZR6Spejh4U02d8jOt6vLEHfe/JQGiRRSQQxSfFWpi1MquVdAyjUar5+76PVCmYl" crossorigin="anonymous"></script>

    <style>
      body {
        /* Stop footer overlapping content*/
          padding-bottom: 120px;
      }

      nav {
        margin-bottom: 25px;
      }
      @media (max-width: 800px) { 
        .container {
          padding: 0px;
          margin-left: 30px;
          margin-right: 30px;
        }
      }

    </style>
</head>

<body>

<link rel="shortcut icon" href="{{ url_for('static', filename='favicon.ico') }}">

<nav class="navbar navbar-dark bg-primary">
  <a class="navbar-brand" href="{{ url_for('index') }}">SHOOT.bio: sprout a branch on the tree of life</a>
</nav>

<div class="container" id = 'Help'>   
    <!-- h1>Help & Documentation</h1 -->   
    <h1>Options:</h1>
       <h3>DIAMOND database: Best respresentatives / All sequences</h3>
              <p class="text-justify">The search against a DIAMOND database is used to assign a gene to the correct homology group  
              prior to phylogenetic placement. The default database (best representatives of homology groups) contains 
              sequences selected using k-means clustering that best represent the sequence diversity of each group. 
              In most cases this works well and the correct homology group for a query gene is easily
              identified. The alternative option is to use the database containing all sequences from all homology
              groups, which may work better in some cases.</p>
       <h3>Initial DIAMOND sensitivity: Default / Ultra-sensitive</h3>
              <p class="text-justify">By default SHOOT performs a DIAMOND normal sensitivity search. If this finds no hits it is followed by a slower 
              "ultra-sensitive" search against the database of all sequences. The alternative option is to perform the slower ultra-sensitive search immediately.</p>
       <h3>MAFFT Options</h3>
              <p class="text-justify">The alignment of the query sequence against the MSA of all the genes in its homology group is one 
              of the most computationally expensive steps in a SHOOT query. For this reason MAFFT is used with options
              giving high performance: "--retree 1 --maxiterate 0 --nofft". You can instead choose for MAFFT to run 
              using its default options, which is slower but may produce more accurate results.</p>
       <h3>Large Trees</h3>
              <p class="text-justify">The largest gene trees in the databases (>2500 genes) have been split into subtrees since phylogentic placement 
              into these large trees can be very slow. At most this affects the 40 largest trees in any database. For these largest trees DIAMOND is
              used to assign the query sequence to a sub-tree and phylogentic methods (MSA & tree inference) are used to place the gene in its correct 
              position wihtin the sub-tree. The sub-tree is then grafted back into the original super-tree, which is returned to the user, thus giving 
              a view of the gene in the complete gene family at lower computational cost. There is, however, a risk of an incorrect placement using this 
              method because a DIAMOND search is not as accurate. In the worst case the best hit could potentially be in a different sub-tree within this 
              large homologous family of genes. In this case, the subsequent phylogentic analysis will be incorrect because an attempt has been made to
              place the gene in the wrong sub-tree. The option 'Phylogenetics on full tree' uses DIAMOND to assign the sequence only to the level of the 
              full tree, and phylogenetic methods are used to determine its position within the tree. While this method it less risky, it could take 3-30 
              minutes or more.</p>
</div>

<footer class="footer fixed-bottom bg-light">
  <div class="container">
    <span class="text-muted">SHOOT.bio: phylogenetic gene search and ortholog inference. David Emms & Steven Kelly &copy; 2021.</span>
  </div>
    <br />
  <div class="container">
    <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">
      <img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/80x15.png" />
    </a>
    <!--br /-->
    This work is licensed under a 
    <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.
  </div>
</footer>

</body>
